mewarpx/mewarpx/diags_store/particle_histogram_diagnostic.py

```diff
@@ -145,37 +109,126 @@ def get_fileprefix(self, species):
+        self.linres = [linres_x, linres_y]
+
+        # for now we hard code this diagnostic to only save weight but this
+        # can be extended in the future
+        self.quantities = ['x', 'y', 'w']
+
+        super(ZplanePHistDiag, self).__init__(
+            diag_steps=diag_steps, species_list=species_list, **kwargs
+        )
 
-        if self.plot:
-            self.plot_histograms()
+    def _process_scraped_particles(self, scraped_particle_dict):
+        """Function to process the scraped particle data.
 
+        Arguments:
+            scraped_particle_dict (dict): Dictionary containing scraped
+                particle data.
+        """
         for ii, species in enumerate(self.species_list):
-            fileprefix = self.get_fileprefix(species.name)
-            np.save(fileprefix + '.npy', self.Harray[ii,...])
+            idx = np.where(
+                scraped_particle_dict['species_id'] == species.species_number
+            )
+            sample = np.zeros((np.size(idx), 2))
+            sample[:,0] = scraped_particle_dict['x'][idx]
+            sample[:,1] = scraped_particle_dict['y'][idx]
+            H, _ = np.histogramdd(
+                sample=sample, bins=self.bins,
+                weights=scraped_particle_dict['w'][idx]
+            )
+            self.Harray[ii,:,:,0] += H
+
+        # if this is a diagnostic period save the data
+        if self.check_timestep():
+            self.save_and_reset_histogram()
+
+    def save_and_reset_histogram(self):
+        """Save and reset the histogram."""
+        # sum the histograms from all the processors
+        self.Harray = parallel_util.parallelsum(self.Harray)
 
+        if mwxrun.me == 0:
+            if self.plot:
+                self.plot_histograms()
+
+            for ii, species in enumerate(self.species_list):
+                fileprefix = self.get_fileprefix(species.name)
+                np.save(fileprefix + '.npy', self.Harray[ii,...])
         self.Harray[:] = 0.0
 
     def plot_histograms(self):
         """Function to plot histogram of scraped particle positions. Special
         plotting functionality is implemented for common assembly types."""
         for ii, species in enumerate(self.species_list):
+            # skip plotting if no data is present
+            if np.all(self.Harray[ii,...] == 0):
+                continue
             plt.title(f"{species.name} scraped on {self.assembly.name}")
-            # plotting for ZPlane assemblies
-            if isinstance(self.assembly, assemblies.ZPlane):
+
+            if mwxrun.dim == 2:
                 plt.xlabel("x position (mm)")
                 plt.ylabel("Current density (A/cm$^2$)")
+
                 dx = (self.bins[0][1] - self.bins[0][0])*1e3
                 data = (
-                    self.Harray[ii,:,0,0,0] * abs(species.sq) / (dx*10)
+                    self.Harray[ii,:,0,0] * abs(species.sq) / (dx*10)
```

Comment:
And what is the `*10` here for?
```suggestion
                    self.Harray[ii,:,0] * abs(species.sq) / (dx*10)
```